libcxx/include/support/atomic/atomic_cuda.h

```diff
@@ -0,0 +1,488 @@
+    NV_DISPATCH_TARGET(
+        NV_IS_DEVICE, (
+            return detail::__atomic_fetch_and_cuda(__a->__get_device(), __pattern, __order, detail::__scope_tag<_Sco>());
+        ),
+        NV_IS_HOST, (
+            return host::__cxx_atomic_fetch_and(__a->__get_host(), __pattern, __order);
+        )
+    )
+}
+
+template <typename _Tp, int _Sco, bool _Ref>
+__host__ __device__
+ _Tp __cxx_atomic_fetch_or(__cxx_atomic_base_heterogeneous_impl<_Tp, _Sco, _Ref> volatile* __a, _Tp __pattern, memory_order __order) {
+    NV_DISPATCH_TARGET(
+        NV_IS_DEVICE, (
+            return detail::__atomic_fetch_or_cuda(__a->__get_device(), __pattern, __order, detail::__scope_tag<_Sco>());
+        ),
+        NV_IS_HOST, (
+            return host::__cxx_atomic_fetch_or(__a->__get_host(), __pattern, __order);
+        )
+    )
+}
+
+template <typename _Tp, int _Sco, bool _Ref>
+__host__ __device__
+ _Tp __cxx_atomic_fetch_xor(__cxx_atomic_base_heterogeneous_impl<_Tp, _Sco, _Ref> volatile* __a, _Tp __pattern, memory_order __order) {
+    NV_DISPATCH_TARGET(
+        NV_IS_DEVICE, (
+            return detail::__atomic_fetch_xor_cuda(__a->__get_device(), __pattern, __order, detail::__scope_tag<_Sco>());
+        ),
+        NV_IS_HOST, (
+            return host::__cxx_atomic_fetch_xor(__a->__get_host(), __pattern, __order);
+        )
+    )
+}
+
+template<class _Tp>
+__host__ __device__ inline uint32_t __cxx_small_to_32(_Tp __val) {
+    __cxx_small_proxy<_Tp> __temp = 0;
+    memcpy(&__temp, &__val, sizeof(_Tp));
+    return __temp;
+}
+
+template<class _Tp>
+__host__ __device__ inline _Tp __cxx_small_from_32(uint32_t __val) {
+    __cxx_small_proxy<_Tp> __temp = static_cast<__cxx_small_proxy<_Tp>>(__val);
+    _Tp __result;
+    memcpy(&__result, &__temp, sizeof(_Tp));
+    return __result;
+}
+
+template <typename _Tp, int _Sco>
+__host__ __device__ inline void __cxx_atomic_init(__cxx_atomic_base_small_impl<_Tp, _Sco> volatile* __a, _Tp __val) {
+    __cxx_atomic_init(&__a->__a_value, __cxx_small_to_32(__val));
+}
+
+template <typename _Tp, int _Sco>
+__host__ __device__ inline void __cxx_atomic_store(__cxx_atomic_base_small_impl<_Tp, _Sco> volatile* __a, _Tp __val, memory_order __order) {
+    __cxx_atomic_store(&__a->__a_value, __cxx_small_to_32(__val), __order);
+}
+
+template <typename _Tp, int _Sco>
+__host__ __device__ inline _Tp __cxx_atomic_load(__cxx_atomic_base_small_impl<_Tp, _Sco> const volatile* __a, memory_order __order) {
+    return __cxx_small_from_32<_Tp>(__cxx_atomic_load(&__a->__a_value, __order));
+}
+
+template <typename _Tp, int _Sco>
+__host__ __device__ inline _Tp __cxx_atomic_exchange(__cxx_atomic_base_small_impl<_Tp, _Sco> volatile* __a, _Tp __value, memory_order __order) {
+    return __cxx_small_from_32<_Tp>(__cxx_atomic_exchange(&__a->__a_value, __cxx_small_to_32(__value), __order));
+}
+__host__ __device__
+inline int __cuda_memcmp(void const * __lhs, void const * __rhs, size_t __count) {
+#ifdef __CUDA_ARCH__
```

Comment:
Do we want to give this `if target` treatment? If not, why?

Reply:
This should be fixed now.